simulation/wallets.js

```diff
@@ -0,0 +1,516 @@
+function unstakeOperators(day) {
+  const unstakingOperatorsToday = unstakingOperators[day]
+  if (!!unstakingOperatorsToday) {
+    unstakingOperatorsToday.forEach((operator) => {
+      delete liveOperators[operator]
+    })
+    delete unstakingOperators[day]
+  }
+}
+
+// Add a poisson random amount of new operators to the network
+function registerNewOperators() {
+  const newOperators = newOperatorCount()
+  for (let j = 0; j < newOperators; j++) {
+    liveOperators[operatorIndex] = true
+    stakingOperators[operatorIndex] = true
+    operatorIndex++
+  }
+}
+
+// Create a new wallet every WALLET_CREATION_PERIOD days
+// FIXME: The actual wallet creation conditions are more complicated (min btc
+// and whatnot), so that's something to implement in the future.
+function createNewWalletEvent(day) {
+  if (day % WALLET_CREATION_PERIOD == 0) {
+    newWallet()
+  }
+}
+
+// Shuffle the staking operators and select 100 to form the signing group.
+function newWallet() {
+  walletBalances[walletIndex] = 0
+  const operators = getRandomSample(Object.keys(stakingOperators), WALLET_SIZE)
+  walletLiveOperators[walletIndex] = {}
+  walletStakingOperators[walletIndex] = {}
+  operators.forEach((operator) => {
+    walletLiveOperators[walletIndex][operator] = true
+    walletStakingOperators[walletIndex][operator] = true
+    let wallets = operatorToWallets[operator] || {}
+    wallets[walletIndex] = true
+    operatorToWallets[operator] = wallets
+  })
+  log(1, "creating new wallet index: " + walletIndex)
+
+  walletIndex++
+}
+
+// An implementation of wallet closure that transfers to a random live wallet.
+function randomTransferWithoutCap(wallet) {
+  let liveWallets = []
+  for (let i = 0; i < walletIndex; i++) {
+    if (
+      walletBalances[i] > WALLET_MIN_BTC &&
+      Object.keys(walletStakingOperators[i]).length >= HEARTBEAT &&
+      i != walletIndex
+    ) {
+      liveWallets.push(i)
+    }
+  }
+  const randomIndex = Math.floor(Math.random() * liveWallets.length)
+  const randomWallet = liveWallets[randomIndex]
+  log(
+    1,
+    "Transferring " +
+      walletBalances[wallet] +
+      " btc from Wallet#" +
+      wallet +
+      " to Wallet#" +
+      randomIndex
+  )
+  walletBalances[randomWallet] += walletBalances[wallet]
+  if (walletBalances[randomWallet] > biggestWalletBalance) {
+    biggestWalletBalance = walletBalances[randomWallet]
+  }
+  numberOfTransfers++
+}
+
+// An implementation of wallet closure that transfers to random live wallet(s)
+// sending out batches of `WALLET_TRANSFER_MAX` before picking a new wallet. If we
+// run out of wallets we start over.
+function randomTransfer(wallet) {
+  let liveWallets = []
+  for (let i = 0; i < walletIndex; i++) {
+    if (
+      walletBalances[i] > WALLET_MIN_BTC &&
+      Object.keys(walletStakingOperators[i]).length >= HEARTBEAT &&
+      i != walletIndex
+    ) {
+      liveWallets.push(i)
+    }
+  }
```

Comment:
If no wallets will be added to `liveWallets` because none will qualify, then rest of the `randomTransfer(wallet)` function will not behave properly - we may end up with situations like this: `Transferring 14 btc from Wallet#0 to Wallet#undefined`.
I haven't tested it, but I think that we may have similar problems when `transferCount` is bigger than length of `liveWallets`.
Maybe we could add a condition that checks if there is enough live wallets, and if not, prints info and cancels the transfer? If we had that, I think condition `Object.keys(walletBalances).length > 1` which I mentioned in comment to `closeWallet` function would be obsolete.

Reply:
BTW, what do you think we should do in real life if the `transferCount` is bigger than the number of live wallets? Distribute the funds from closing wallet evenly between live wallets?

Reply:
We're okay in the *current* version of the sim because the newest wallet will never close and will always be a candidate. If I ever let that close, then it breaks some assumptions and I'll have to go back and rewrite some other functions (like this one) to handle it. Hopefully that's fine in a follow-up!

As for `transferCount`, the new implementation in [ffa3b24](https://github.com/keep-network/tbtc-v2/pull/168/commits/ffa3b247b8b290ae3c1fc8e52b382a03d4abdc8a) handles that by defining it as:

```
const transferCount = Math.min(liveWallets.length, Math.ceil(walletBalances[wallet] / WALLET_TRANSFER_MAX))
```
so that transfer count will never exceed the number of live wallets!